oximeter/db/src/sql/mod.rs

```diff
@@ -0,0 +1,1351 @@
+            qualify: None,
+        }
+    }
+
+    // Process a single "table factor", the <object> in `FROM <object>` to
+    // extract the names of the timeseries it refers to.
+    //
+    // Note this is recursive since we do support basic inner joins.
+    fn process_table_factor(
+        relation: &mut TableFactor,
+    ) -> Result<IndexSet<TimeseriesName>, OxdbError> {
+        match relation {
+            TableFactor::Table { ref mut name, args, with_hints, .. } => {
+                if args.is_some() || !with_hints.is_empty() {
+                    return unsupported!(
+                        "Table functions and hints are not supported"
+                    );
+                }
+                if name.0.len() != 1 {
+                    return unsupported!(
+                        "Query must select from single named \
+                        table, with no database"
+                    );
+                }
+                let timeseries_name = name.0[0]
+                    .value
+                    .parse()
+                    .map(|n| indexmap::indexset! { n })
+                    .map_err(|_| MetricsError::InvalidTimeseriesName)?;
+                // Rewrite the quote style to be backticks, so that the
+                // resulting actual query translates into a valid identifier for
+                // ClickHouse, naming the CTE's well generate later.
+                name.0[0].quote_style = Some('"');
+                Ok(timeseries_name)
+            }
+            TableFactor::Derived { lateral: false, subquery, .. } => {
+                RestrictedQuery::process_query(subquery)
+            }
+            _ => {
+                return unsupported!(
+                "Query must select from concrete tables or subqueries on them"
+            )
+            }
+        }
+    }
+
+    // Process a parsed query, returning the named timeseries that it refers to.
+    //
+    // This is the entry-point for our query processing implementation. We take
+    // a parsed query from `sqlparser`, and extract the virtual tables
+    // (timeseries names) that we'll need to construct in order to actually run
+    // it against our database.
+    //
+    // Note that we return an _ordered set_ of the timeseries names. This is to
+    // produce the CTEs that correspond to each timeseries, but without
+    // duplicating the actual CTE.
+    fn process_query(
+        query: &mut Query,
+    ) -> Result<IndexSet<TimeseriesName>, OxdbError> {
+        // Some basic checks limiting the scope of the query.
+        if query.with.is_some()
+            || query.fetch.is_some()
+            || !query.locks.is_empty()
+        {
+            return unsupported!(
+                "CTEs, FETCH and LOCKS are not currently supported"
+            );
+        }
+        let SetExpr::Select(select) = &mut *query.body else {
+            return unsupported!("Only SELECT queries are currently supported");
+        };
```

Comment:
was not aware of let/else. that is handy!

Reply:
Yeah, might be my favorite syntax enhancement from the last few Rust releases!